docs/syntax-key-destructuring.md

```diff
@@ -0,0 +1,182 @@
+# Key destructuring
+
+## Summary
+
+Introduce a new syntax for unpacking key values into their own variables, such that:
+
+```lua
+local { .a, .b } = t
+-- a == t.a
+-- b == t.b
+```
+
+## Motivation
+
+Simple indexes on tables are very common both in and outside of Luau. A common use case is large libraries. It is common in the web world to see something like:
+
+```js
+const { useState, useEffect } = require("react");
+```
+
+...which allows you to quickly use `useState` and `useEffect` without fully qualifying it in the form of `React.useState` and `React.useEffect`. In Luau, if you do not want to fully qualify common React functions, the top of your file will often look like:
+
+```lua
+local useEffect = React.useEffect
+local useMemo = React.useMemo
+local useState = React.useState
+-- etc
+```
+
+...which creates a lot of redundant cruft.
+
+It is also common to want to have short identifiers to React properties, which basically always map onto a variable of the same name. As an anecdote, a regex search of `^\s+local (\w+) = \w+\.\1$` comes up 103 times in the My Movie codebase, many in the form of indexing React properties:
+
+```lua
+local position = props.position
+local style = props.style
+-- etc...
+```
+
+...whereas in JavaScript this would look like:
+```js
+const { position, style } = props
+
+// Or even...
+function MyComponent({
+	position,
+	style,
+})
+```
+
+React properties are themselves an example of a common idiom of passing around large tables as function arguments, such as with HTTP requests:
+
+```js
+// JavaScript
+get("/users", ({
+	users,
+	nextPageCursor,
+}) => { /* code */ })
+```
+
+## Design
+
+This RFC proposes expanding the grammar of `binding`:
+```patch
+binding = NAME [':' Type]
++         | `{` keydestructor [`,` keydestructor] [`,`] `}` [':' Type]
+
++keydestructor = `.` NAME [':' Type]
+```
+
+This would allow for the following:
+
+```lua
+local { .a, .b }, c = t
+
+for _, { .a, .b } in ts do
+end
+
+local function f({ .a, .b }, c)
+end
```

Comment:
Can we please drop this?
Not sure why it's even proposed (argument list bindings are a whole different construct from local bindings)
It increases complexity of the implementation and I don't think this had strong motivation for it (in many discussions, discussion always came up in context of locals, not argument lists).
This introduces unnamed function arguments.
It also raises questions of how many arguments the function really has. `local {.a, .b}` creates two locals, but `f({.a, .b})` is one argument.

Since overall the value of the proposal is already low (as in, just do local t = smth; just do local a, b = t.a, t.b; it's not that hard and you can enjoy all the renaming you need), the complexity of this moves the proposal further into the area of negative points it has going for itself.

Reply:
I think I agree on some fronts. This makes people tempted to author functions that requires the user to heap-allocate a table only to throw it away after the function call ends.

```lua
local function say({.name, .message})
  print(`{name}: {message}`)
end

-- heap allocate a table and then discarded.
-- garbage collector is not having a fun day!
say{
  name = "Alex",
  message = "Hello, world!"
}
```

That said, I don't resonate with some of the counterarguments...

> This introduces unnamed function arguments.

I'm not sure this is a strong argument against.

> It also raises questions of how many arguments the function really has. local {.a, .b} creates two locals, but f({.a, .b}) is one argument.

I think it's obvious: 1 argument. It just is not named.

Reply:
> I'm not sure this is a strong argument against. Nobody bats an eye in C++, JavaScript, and even Rust (with irrefutable patterns)

It's a new concept for Luau compiler and debugger.
And this feature is not that great to have this going for it.

Reply:
And since we mentioned it, C++ had good taste to not include destructuring in argument list.

Reply:
I'm fine with it

Reply:
Another possible argument against it: there is no place in the syntax to do key destructuring on
 `self` despite that it's still as useful too. I just wrote a function today that only uses `ZoomLevel` from `self`.

```lua
function Camera:GetCameraOffset() -- Where to say `{.ZoomLevel}` as `self`?
    local {.ZoomLevel} = self
    return 
end
```

Reply:
Just want to note that we can always propose additions in the future to extend to function arguments.

We can see how developers will start using this and if it's often used to destructure function argument at the start of the function, that will be a good signal to consider how to make it possible.

Even for loop can be added later, but I think we'll be ok with it now (even though that requires separate implementation work from local statement).